metadata-io/src/main/java/com/linkedin/metadata/search/elasticsearch/query/request/SearchQueryBuilder.java

```diff
@@ -279,23 +280,35 @@ private static FunctionScoreQueryBuilder.FilterFunctionBuilder[] buildAnnotation
     finalScoreFunctions.add(
             new FunctionScoreQueryBuilder.FilterFunctionBuilder(ScoreFunctionBuilders.weightFactorFunction(1.0f)));
 
-    entitySpecs.stream()
+    Map<String, List<SearchableAnnotation>> annotations = entitySpecs.stream()
         .map(EntitySpec::getSearchableFieldSpecs)
         .flatMap(List::stream)
         .map(SearchableFieldSpec::getSearchableAnnotation)
-        .flatMap(annotation -> annotation
-            .getWeightsPerFieldValue()
-            .entrySet()
-            .stream()
-            .map(entry -> buildWeightFactorFunction(annotation.getFieldName(), entry.getKey(),
-                entry.getValue())))
-        .forEach(finalScoreFunctions::add);
+        .collect(Collectors.groupingBy(SearchableAnnotation::getFieldName,
+        Collectors.mapping(annotation -> annotation,
+            Collectors.toList())));
+
+    for (Map.Entry<String, List<SearchableAnnotation>> annotationEntry : annotations.entrySet()) {
```

Comment:
Hmm, can't this just be changed in the lambda that's currently in the flatMap call? As in instead of streaming across all of them just pull the first one and calculate the weight factor? Doing an intermediate collect complicates the readability and is relatively expensive.

Reply:
To deduplicate equivalent values in the stream you can use distinct()

Reply:
Yeah, I can do that! This is how it's done above but I'll rewrite it to do that (and maybe the one above too)